Capture uniform buffers

ilkkasaa opened the pull request:
Adds functionality to track descriptor set bindings and uniform buffers.
Stores the captured uniform buffers to binary files.
Also contains preparation to support other descriptor types.

ilkkasaa committed:
Capture uniform buffers

Adds functionality to track descriptor set bindings and uniform buffers.
Stores the captured uniform buffers to binary files.
Also contains preparation to support other descriptor types.

ilkkasaa committed:
Fix "not all control paths return a value"

ilkkasaa committed:
Fix typo

ilkkasaa committed:
Reverse DESCRIPTOR_RANGE and DESCRIPTOR_OFFSET

ilkkasaa committed:
Fix wrong descriptor range in some cases

Descriptor range was computed wrongly if dynamic offsets were used and
descriptor write contained VkDescriptorBufferInfo with range of
VK_WHOLE_SIZE.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/descriptor_set_data.h:
```diff
@@ -0,0 +1,86 @@
+  }*/
+
+ private:
+  // Layout used to allocate this descriptor set.
+  const DescriptorSetLayoutData* descriptor_set_layout_data_;
```
So it looks like `DescriptorSetData` does not own the `DescriptorSetLayoutData`. I think this is almost OK because the Vulkan spec says you must not update a descriptor set after destroying the layout. However, you also have a getter that is used when processing a draw command; I believe the layout can be destroyed by this point. Thus, I think we will have to make a deep copy of the layout data and `DescriptorSetData` will own that copy. I could be wrong though.

ilkkasaa replied:
Good catch! I need to do some investigation if the layout can be destroyed when the draw command is being processed.

ilkkasaa replied:
Yeah, you're right! The layout can be destroyed by that point. I'll change it to make a copy of the `DescriptorSetData`. I guess we could also use shared ownership here, but I think the performance gain achieved by avoiding deep copy of one class and a struct isn't good enough reason to use shared_ptr.

paulthomson reviewed (commented):
Awesome! I left one comment for now.

ilkkasaa committed:
Fix descriptor set layout lifetime issue

ilkkasaa committed:
Add missing include

ilkkasaa committed:
Github actions: remove builds using old compilers

Removes github actions using Clang and GCC version < 9.x.
Uses Ubuntu 18.04 instead of 16.04.

ilkkasaa committed:
Reverse dynamic offset and descriptor offset and range

Reverse the order of dynamic offset OFFSET and DESCRIPTOR_OFFSET
and DESCRIPTOR_RANGE to match the Amber script syntax.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/descriptor_set_data.h:
```diff
@@ -0,0 +1,87 @@
+  // descriptors will be stored when updating the descriptor set.
+  explicit DescriptorSetData(const DescriptorSetLayoutData* layout_data);
+
+  // Get the descriptor bindings of buffer type.
```
Google C++ style says that doc comments should start "Returns ..." or "Gets...". Usually, "Returns..." works well. Same thing below.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/descriptor_set_data.h:
```diff
@@ -0,0 +1,87 @@
+  /*
+  void CopyDescriptorSet(const VkCopyDescriptorSet& copy_descriptor_set) {
+    RUNTIME_ASSERT_MSG(false, "CopyDescriptorSet not implemented.");
+  }*/
```
Can we remove? Leaving a TODO is fine, but we prefer not to leave commented out code.

paulthomson commented on src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/draw_call_tracker.h:
```diff
@@ -33,6 +37,20 @@ namespace gf_layers::amber_scoop_layer {
+    // the descriptor(s) and value is a list of dynamic offset(s) for the
+    // binding.
+    std::unordered_map<BindingNumber, std::vector<BindingNumber>>
+        dynamic_offsets;
```
Shouldn't the vector type parameter be DynamicOffset?

paulthomson commented on src/VkLayer_GF_amber_scoop/src/amber_scoop_layer.cc:
```diff
@@ -389,6 +493,38 @@ VKAPI_ATTR VkResult VKAPI_CALL vkCreateGraphicsPipelines(
 }
 
+//
+// Our vkCreateShaderModule function.
```
Typo. vkCreateShaderModule -> vkCreatePipelineLayout

paulthomson commented on src/VkLayer_GF_amber_scoop/src/vk_deep_copy.cc:
```diff
@@ -32,6 +32,51 @@ void DeepDelete(const VkBufferCreateInfo& create_info) {
+    const VkDescriptorSetLayoutCreateInfo& create_info) {
+  VkDescriptorSetLayoutCreateInfo result = create_info;
+  // NOLINTNEXTLINE(cppcoreguidelines-owning-memory)
+  auto* bindings = new VkDescriptorSetLayoutBinding[create_info.bindingCount];
```
I feel like bindingCount could be 0. Unlikely though!

paulthomson reviewed (commented):
Really nice! A few minor comments.

ilkkasaa committed:
Minor fixes

ilkkasaa committed:
Fix gcc build error

Try to fix clang build error by updating cppcheck from 2.1 to 2.4.1

ilkkasaa committed:
Fix cppcheck build error with clang-9

Checkout specific commit of cppcheck to fix the build errors.
